python/ray/_private/structured_logging/logging_config.py

```diff
@@ -0,0 +1,52 @@
+from typing import Optional
+
+from ray._private.structured_logging.constants import LOG_MODE_DICT
+from ray.util.annotations import PublicAPI
+
+
+@PublicAPI(stability="alpha")
+class LoggingConfig:
```

Comment:
Actually I think we can make it a dataclass. Other configs in the codebase are dataclass such as `FailureConfig`

Reply:
What's the benefit of using `dataclass` here? It is not obvious to me, especially since we want to verify whether `encoding` is in `LOG_MODE_DICT` and raise a `ValueError` if it doesn't exist in the `__init__` function. Hence, we can't benefit from the automatic generation of `__init__` by `dataclass`.

Reply:
dataclass has `__post_init__()` for validation purpose.

Reply:
```
Python's dataclass is a decorator that provides an easy way to create classes that primarily store data without having to write boilerplate code. Here are some key advantages of using dataclass:

Advantages of Using dataclass
Reduced Boilerplate Code:

Automatically generates __init__, __repr__, __eq__, and other special methods based on the class attributes.
Saves time and effort by reducing the amount of code you need to write and maintain.
Improved Readability:

Enhances code readability and clarity by explicitly defining data structures in a concise manner.
Makes it clear that the class is intended to be used as a simple data container.
Built-in Methods:

Provides default implementations for common methods:
__init__: Initializes the class with the given attributes.
__repr__: Provides a string representation of the class instance, useful for debugging.
__eq__: Compares instances of the class based on attribute values.
__hash__: Allows instances to be used in hashed collections (e.g., sets, dictionaries) if frozen=True is set.
Type Annotations:

Encourages the use of type annotations, improving code quality and aiding in static analysis and IDE support.
Mutability Control:

Allows control over mutability with the frozen parameter:
frozen=True makes the instances of the class immutable, similar to tuples.
Ensures data integrity by preventing modification of attributes.
Default Values and Factory Functions:

Supports default values and default factory functions for attributes, providing flexibility in instance creation.
Automatic Order Methods:

With the order=True parameter, automatically generates comparison methods (__lt__, __le__, __gt__, __ge__) based on the order of attributes.
Easy Conversion to/from Other Formats:

Simplifies conversion between data classes and other formats (e.g., dictionaries, JSON) using utility functions.
```

Told by chatgpt

Reply:
> dataclass has __post_init__() for validation purpose.

I knew it. That's why I said that "we can't benefit from the automatic generation of __init__ by dataclass". I also asked ChatGPT and read something similar to what you pasted above, but didn't find the benefits we were looking for in this case. I will update it to `dataclass` to keep it consistent with other config classes, but I am still curious about the motivation.

Reply:
Updated https://github.com/ray-project/ray/pull/45344/commits/5083efb8f37729b7609ca44dfd074c1d228b264e